tests/php/Forms/FormFieldTest.php

```diff
@@ -455,6 +457,35 @@ public function testGetSchemaStateWithFormValidation()
         );
     }
 
+    public function testValidationExtensionHooks()
+    {
+        /** @var TextField|FieldValidationExtension $field */
+        $field = new TextField('Test');
```

Comment:
Is there an easy way to loop through `ClassInfo::subclassesFor(FormField::class)` and run through this test with all our formfields? That would be a good way to ensure:
1, we haven't missed any this time around (there will be some in other modules we'll have to come back and add this to, for example, and running this test against the sink would catch them)
2. any new form fields have this functionality applied

Reply:
I’ve been trying to achieve this but I can’t do it nicely. A lot of the subclasses have their own constructor argument requirements, and while I can hardcode specific arguments for specific tests, it’s going to make the test quite fragile (and long!). This is the best I could come up with, do you think it’s worth adding?

```php
    public function testValidationExtensionHooksAreCalledOnFormFieldSubclasses()
    {
        // Can't use a provider for this due to class manifest weirdness
        $formFieldClasses = ClassInfo::subclassesFor(FormField::class, false);
        foreach ($formFieldClasses as $formFieldClass) {
            // Skip over anything GridField-related
            if (strpos($formFieldClass, 'GridField') !== false) {
                continue;
            }

            $reflection = new ReflectionClass($formFieldClass);
            // Skip abstract classes like MultiSelectField
            if ($reflection->isAbstract()) {
                continue;
            }

            /** @var FormField $field */
            $field = $reflection->newInstanceWithoutConstructor();
            // Skip readonly and data-less fields
            if (!$field->hasData() || $field->isReadonly()) {
                continue;
            }

            $args = ['Test'];
            if ($formFieldClass === NullableField::class) {
                $args = [TextField::create('Test2')];
            } else if ($formFieldClass === TreeDropdownField::class) {
                $args = ['Test', 'Test', Group::class];
            } else if ($formFieldClass === PermissionCheckboxSetField::class) {
                $args = ['Test', 'Test', Permission::class, 'Test'];
            } else if ($formFieldClass === SelectionGroup::class) {
                $args = ['Test', []];
            }

            $mock = $this->getMockBuilder($formFieldClass)
                ->setConstructorArgs($args)
                ->setMethods(['extendValidationResult'])
                ->getMock();

            $mock->expects($this->once())
                ->method('extendValidationResult')
                ->will($this->returnValue(true));

            $this->assertTrue($mock->validate(new RequiredFields()));
        }
    }
```

Reply:
> // Skip over anything GridField-related

Does this not work with gridfields?

> // Skip readonly and data-less fields

I'm guessing `validate()` doesn't get called on these? That would make sense, since the user can't set values for them.

> This is the best I could come up with, do you think it’s worth adding?

Yes, definitely. This will catch any scenario where we add new fields but forget about the new validation extending. Worst case we need to add another "else if" (maybe make it a switch but I'm fine with it as is for now) when adding a new field. It'll also catch if we break this with any given field.

Well, worst case we do something really funky and the new field doesn't work with this, but we can cross that bridge if we come to it.

Reply:
>> // Skip over anything GridField-related

>Does this not work with gridfields?

Yes, but you can’t really “validate” a GridField anyway. The main reason for skipping was to reduce the number of "elseif" blocks, but I’ll include them for now and we can see how long it is and whether we want to exclude any

>> // Skip readonly and data-less fields

> I'm guessing validate() doesn't get called on these? That would make sense, since the user can't set values for them.

It does, again I just skipped them to reduce the number of elseif blocks. As above I’ll add everything for now and see how long it ends up

Reply:
Okay I’ve pushed that test, it didn’t end up as long as I’d feared 😄